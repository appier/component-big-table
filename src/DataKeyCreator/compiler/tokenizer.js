
const WHITESPACE_TEST = /\s/;
const NUMBERS_TEST = /[0-9]/;
const OPERATOR_TEST = /[\-+*/]/;
const IDENTIFIER_HEAD_TEST = /[a-z]/i;
const IDENTIFIER_TEST = /[a-z0-9_\-]/i;

const tokenizer = (input) => {

  let position = 0;
  let tokens = [];

  while(position < input.length){
    let char = input[position];

    if (char === '(') {
      tokens.push({
        type: 'paren',
        value: '(',
      });
      position++;
      continue;
    }

    if (char === ')') {
      tokens.push({
        type: 'paren',
        value: ')',
      });
      position++;
      continue;
    }

    //ignore white space..
    if (WHITESPACE_TEST.test(char)) {
      position++;
      continue;
    }

    //abstract identifier
    if (IDENTIFIER_HEAD_TEST.test(char)) {
      let value = '';
      while (
        (IDENTIFIER_TEST.test(char) || NUMBERS_TEST.test(char))
        && position < input.length
      ) {
        value += char;
        char = input[++position];
      }
      tokens.push({
        type: 'identifier',
        value: value
      });
      continue;
    }

    //abstract number
    if(NUMBERS_TEST.test(char)) {
      let value = '';
      while (NUMBERS_TEST.test(char) && position < input.length) {
        value += char;
        char = input[++position];
      }
      tokens.push({
        type: 'number',
        value: value
      });
      continue;
    }

    //abstract operator
    if(OPERATOR_TEST.test(char)) {
      let value = '';
      while (OPERATOR_TEST.test(char) && position < input.length) {
        value += char;
        char = input[++position];
        if(value.length >= 2){
          throw new TypeError('illegal operator: '+ value);
        }
      }
      tokens.push({
        type: 'operator',
        value: value
      });
      continue;
    }


    throw new TypeError('Unable to parse character: ' + char);
  }

  return tokens;
};

export default tokenizer;