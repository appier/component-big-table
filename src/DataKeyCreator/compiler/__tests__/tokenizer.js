import tokenizer from '../tokenizer';

describe('Tokenizer', () => {

  it('identifier', () => {
    expect(tokenizer('cost')).toMatchSnapshot();
  });

  it('identifier with number', () => {
    expect(tokenizer('action1 / action2')).toMatchSnapshot();
  });

  it('operator', () => {
    expect(tokenizer('+ - * /')).toMatchSnapshot();
  });

  it('paren', () => {
    expect(tokenizer(')()')).toMatchSnapshot();
  });

  it('number', () => {
    expect(tokenizer('1 23 45')).toMatchSnapshot();
  });

  it('identifier & operator', () => {
    expect(tokenizer('cost / action')).toMatchSnapshot();
  });

  it('paren & identifier', () => {
    expect(tokenizer('cost ( action)')).toMatchSnapshot();
  });

  it('number & operator', () => {
    expect(tokenizer(' 1 + 2 * 3 - 4 / 5')).toMatchSnapshot();
  });

  it('identifier & number & operator', () => {
    expect(tokenizer(' impression * 1000 / cost ')).toMatchSnapshot();
  });

  it('identifier & paren & number & identifier', () => {
    expect(tokenizer('(impression + click)/(cost*5)')).toMatchSnapshot();
  });

  it('operator: invalid 1', () => {
    try {
      tokenizer('++')
    } catch (err) {
      expect(String(err)).toMatchSnapshot();
    }
  });

  it('operator: invalid 2', () => {
    try {
      tokenizer('**')
    } catch (err) {
      expect(String(err)).toMatchSnapshot();
    }
  });


})