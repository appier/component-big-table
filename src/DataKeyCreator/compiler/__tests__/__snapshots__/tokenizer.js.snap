exports[`Tokenizer identifier & number & operator 1`] = `
Array [
  Object {
    "type": "identifier",
    "value": "impression",
  },
  Object {
    "type": "operator",
    "value": "*",
  },
  Object {
    "type": "number",
    "value": "1000",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "identifier",
    "value": "cost",
  },
]
`;

exports[`Tokenizer identifier & operator 1`] = `
Array [
  Object {
    "type": "identifier",
    "value": "cost",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "identifier",
    "value": "action",
  },
]
`;

exports[`Tokenizer identifier & paren & number & identifier 1`] = `
Array [
  Object {
    "type": "paren",
    "value": "(",
  },
  Object {
    "type": "identifier",
    "value": "impression",
  },
  Object {
    "type": "operator",
    "value": "+",
  },
  Object {
    "type": "identifier",
    "value": "click",
  },
  Object {
    "type": "paren",
    "value": ")",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "paren",
    "value": "(",
  },
  Object {
    "type": "identifier",
    "value": "cost",
  },
  Object {
    "type": "operator",
    "value": "*",
  },
  Object {
    "type": "number",
    "value": "5",
  },
  Object {
    "type": "paren",
    "value": ")",
  },
]
`;

exports[`Tokenizer identifier 1`] = `
Array [
  Object {
    "type": "identifier",
    "value": "cost",
  },
]
`;

exports[`Tokenizer identifier with number 1`] = `
Array [
  Object {
    "type": "identifier",
    "value": "action1",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "identifier",
    "value": "action2",
  },
]
`;

exports[`Tokenizer number & operator 1`] = `
Array [
  Object {
    "type": "number",
    "value": "1",
  },
  Object {
    "type": "operator",
    "value": "+",
  },
  Object {
    "type": "number",
    "value": "2",
  },
  Object {
    "type": "operator",
    "value": "*",
  },
  Object {
    "type": "number",
    "value": "3",
  },
  Object {
    "type": "operator",
    "value": "-",
  },
  Object {
    "type": "number",
    "value": "4",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "number",
    "value": "5",
  },
]
`;

exports[`Tokenizer number 1`] = `
Array [
  Object {
    "type": "number",
    "value": "1",
  },
  Object {
    "type": "number",
    "value": "23",
  },
  Object {
    "type": "number",
    "value": "45",
  },
]
`;

exports[`Tokenizer operator 1`] = `
Array [
  Object {
    "type": "operator",
    "value": "+",
  },
  Object {
    "type": "operator",
    "value": "-",
  },
  Object {
    "type": "operator",
    "value": "*",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
]
`;

exports[`Tokenizer operator: invalid 1 1`] = `"TypeError: illegal operator: ++"`;

exports[`Tokenizer operator: invalid 2 1`] = `"TypeError: illegal operator: **"`;

exports[`Tokenizer paren & identifier 1`] = `
Array [
  Object {
    "type": "identifier",
    "value": "cost",
  },
  Object {
    "type": "paren",
    "value": "(",
  },
  Object {
    "type": "identifier",
    "value": "action",
  },
  Object {
    "type": "paren",
    "value": ")",
  },
]
`;

exports[`Tokenizer paren 1`] = `
Array [
  Object {
    "type": "paren",
    "value": ")",
  },
  Object {
    "type": "paren",
    "value": "(",
  },
  Object {
    "type": "paren",
    "value": ")",
  },
]
`;
